shapeshift/web
# feat: move KeepKey's pin changing behavior to the wallet menu

## Description
- Moved the `Pin` logic and markup to a reusable component and added some props needed to customize it for the menu
- Fixed a bug crashing the app inside error handling of the new pin action
- Added a new wallet state `isUpdatingPin` in order to prevent the regular modal to open when the reset pin action is launched from the menu
- Added a new wallet state `isLoading` so we can force a loading state when the device is loading for example
- I used polling to handle the passphrase sent event response, because the KeepKey is silent on success, see code comments for more information!

⚠️ There is a keepkey bug when cancelling / going back to previous route, the `cancel` method from hdwallet makes the keepkey crazy, I guess this is a firmware issue and will be fixed with https://github.com/shapeshift/web/issues/1613



## Notice




## Pull Request Type


## Issue (if applicable)



closes #1414

## Risk



- The reset pin feature could be broken or buggy
- The enter pin feature could be broken or buggy in case I did an error while moving to a reusable component

## Testing



- Connect your keepkey
- Open the wallet menu on the top-right and try to reset the pin of the keepkey
- Complete the whole pin reset process, everything should work as expected until the end
- Try to also cancel, go back to the previous menu screen
- You'll have more information in the issue, mockups and as a video in the screenshots part

### Engineering



- Connect your keepkey
- Open the wallet menu on the top-right and try to reset the pin of the keepkey
- Complete the whole pin reset process, everything should work as expected until the end
- Try to also cancel, go back to the previous menu screen
- You'll have more information in the issue, mockups and as a video in the screenshots part

### Operations



- Connect your keepkey
- Open the wallet menu on the top-right and try to reset the pin of the keepkey
- Complete the whole pin reset process, everything should work as expected until the end
- Try to also cancel, go back to the previous menu screen
- You'll have more information in the issue, mockups and as a video in the screenshots part

## Screenshots (if applicable)
https://www.awesomescreenshot.com/video/11006360?key=80a2f395eaa20c19c602b2d95329e46f
